pkg/network/usm/sharedlibraries/watcher_test.go

```diff
@@ -104,40 +106,89 @@ func (s *SharedLibrarySuite) TestSharedLibraryDetection() {
 	t := s.T()
 
-	// Since we want to detect that the write _hasn't_ been detected, verify the
-	// read too to try to ensure that test isn't broken and failing to detect
-	// the write due to some bug in the test itself.
-	readPath, readPathID := createTempTestFile(t, "read-foo-libssl.so")
-	writePath, writePathID := createTempTestFile(t, "write-foo-libssl.so")
-
-	registerRecorder := new(utils.CallbackRecorder)
-	unregisterRecorder := new(utils.CallbackRecorder)
-
-	watcher, err := NewWatcher(utils.NewUSMEmptyConfig(), LibsetCrypto,
-		Rule{
-			Re:           regexp.MustCompile(`foo-libssl.so`),
-			RegisterCB:   registerRecorder.Callback(),
-			UnregisterCB: unregisterRecorder.Callback(),
+	tests := []struct {
+		syscallType string
+		skipFunc    func(t *testing.T)
+	}{
+		{
+			syscallType: "open",
 		},
-	)
-	require.NoError(t, err)
-	watcher.Start()
-	t.Cleanup(watcher.Stop)
-
-	// Use a sleep 1 as in TestSharedLibraryDetectionWithPIDAndRootNamespace
-	// below to give the watcher a chance to detect the process.
-	_, err = exec.Command("sh", "-c",
-		fmt.Sprintf("sleep 1 < %s > %s", readPath, writePath)).CombinedOutput()
-	require.NoError(t, err)
-
-	require.EventuallyWithT(t, func(c *assert.CollectT) {
-		assert.Equal(c, 1, registerRecorder.CallsForPathID(readPathID))
-		assert.Equal(c, 0, registerRecorder.CallsForPathID(writePathID))
-	}, time.Second*5, 10*time.Millisecond)
+		{
+			syscallType: "openat",
+		},
+		{
+			syscallType: "openat2",
+			skipFunc: func(t *testing.T) {
+				if !sysOpenAt2Supported() {
+					t.Skip("openat2 not supported")
+				}
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.syscallType, func(t *testing.T) {
+			if tt.skipFunc != nil {
+				tt.skipFunc(t)
+			}
+			// Since we want to detect that the write _hasn't_ been detected, verify the
+			// read too to try to ensure that test isn't broken and failing to detect
+			// the write due to some bug in the test itself.
+			readPath, readPathID := createTempTestFile(t, "read-foo-libssl.so")
+			writePath, writePathID := createTempTestFile(t, "write-foo-libssl.so")
+
+			registerRecorder := new(utils.CallbackRecorder)
+			unregisterRecorder := new(utils.CallbackRecorder)
+
+			watcher, err := NewWatcher(utils.NewUSMEmptyConfig(), LibsetCrypto,
+				Rule{
+					Re:           regexp.MustCompile(`foo-libssl.so`),
+					RegisterCB:   registerRecorder.Callback(),
+					UnregisterCB: unregisterRecorder.Callback(),
+				},
+			)
+			require.NoError(t, err)
+			watcher.Start()
+			t.Cleanup(watcher.Stop)
+			// Overriding PID, to allow the watcher to watch the test process
+			watcher.thisPID = 0
+
+			how := unix.OpenHow{Mode: 0644}
+
+			require.EventuallyWithT(t, func(c *assert.CollectT) {
+				how.Flags = syscall.O_CREAT | syscall.O_RDONLY
+				fd, err := open(unix.AT_FDCWD, readPath, &how, tt.syscallType)
+				require.NoError(c, err)
+				require.NoError(c, syscall.Close(fd))
+				assert.Equal(c, 1, registerRecorder.CallsForPathID(readPathID))
```

Comment:
Shouldn't this be `assert.GreaterOrEqual`? If the registration is delayed slightly in the first try you could end up with the `CallsForPathID` returning zero in the first try and 2 in the second one.